### PR TITLE
CNF-15742: Add Reinstall spec and status fields to ClusterInstance

### PR DESCRIPTION
**Pull request by sakhoury**
# Summary
This PR introduces the new reinstall spec and status fields to the ClusterInstance CRD.

**Resolves:** [CNF-15742](https://issues.redhat.com/browse/CNF-15742)

### CODE REVIEW

File: api/v1alpha1/clusterinstance_types.go [hunk 1]
```diff
@@ -215,6 +215,31 @@ const (
...
+	// This field informs the SiteConfig Operator whether to preserve data or not.
+	// Furthermore, it provides a mechanism to ensure / enforce that cluster identity data is preserved if
+	// configured to ClusterIdentity.
+	// Possible values:
+	// - None: no data preservation is required.
+	// - All: All Secrets and / or ConfigMaps with the preservation-label are preserved.
+	// - ClusterIdentity: only Secrets and / or ConfigMaps with the cluster-identity preservation-label are preserved.
+	// +kubebuilder:validation:Enum=None;All;ClusterIdentity
+	// +kubebuilder:default=None
+	// + required
```
Comment (carbonin) [inline api/v1alpha1/clusterinstance_types.go:239]: I feel like there shouldn't be a space after the `+`

File: api/v1alpha1/clusterinstance_types.go [hunk 2]
```diff
@@ -406,6 +436,27 @@ type ManifestReference struct {
 	Message string `json:"message,omitempty"`
 }
 
+type ReinstallHistory struct {
+	// Observed reinstall generation.
+	// + required
```
Comment (carbonin) [inline api/v1alpha1/clusterinstance_types.go:441]: Same

File: api/v1alpha1/clusterinstance_types.go [hunk 3]
```diff
@@ -406,6 +436,27 @@ type ManifestReference struct {
...
+	ObservedGeneration string `json:"observedGeneration"`
+	// Date/time of the reinstallation.
+	// +required
+	Timestamp metav1.Time `json:"timestamp"`
+	// Difference between last and current ClusterInstnace spec fields as a JSON string.
+	// + required
```
Comment (carbonin) [inline api/v1alpha1/clusterinstance_types.go:447]: Same

File: api/v1alpha1/clusterinstance_types.go [hunk 4]
```diff
@@ -215,6 +215,31 @@ const (
...
+
+type ReinstallSpec struct {
+	// Desired reinstall generation. Used to trigger a reinstallation.
+	// +required
+	Generation string `json:"generation"`
+	// This field informs the SiteConfig Operator whether to preserve data or not.
+	// Furthermore, it provides a mechanism to ensure / enforce that cluster identity data is preserved if
+	// configured to ClusterIdentity.
+	// Possible values:
+	// - None: no data preservation is required.
```
Comment (carbonin) [inline api/v1alpha1/clusterinstance_types.go:234]: Required or will be performed?

File: api/v1alpha1/clusterinstance_types.go [hunk 5]
```diff
@@ -215,6 +215,31 @@ const (
...
+	// - All: All Secrets and / or ConfigMaps with the preservation-label are preserved.
+	// - ClusterIdentity: only Secrets and / or ConfigMaps with the cluster-identity preservation-label are preserved.
```
Comment (carbonin) [inline api/v1alpha1/clusterinstance_types.go:236]: Can this be more specific about the label key and value?

File: api/v1alpha1/clusterinstance_types.go [hunk 6]
```diff
@@ -406,6 +436,27 @@ type ManifestReference struct {
...
 }
 
+type ReinstallHistory struct {
+	// Observed reinstall generation.
+	// + required
+	ObservedGeneration string `json:"observedGeneration"`
+	// Date/time of the reinstallation.
+	// +required
+	Timestamp metav1.Time `json:"timestamp"`
+	// Difference between last and current ClusterInstnace spec fields as a JSON string.
```
Comment (carbonin) [inline api/v1alpha1/clusterinstance_types.go:446]: This is interesting. How large do we expect this to be?

File: api/v1alpha1/clusterinstance_types.go (hunk 1 above)
Comment (sakhoury) [reply api/v1alpha1/clusterinstance_types.go:239]: 🤦‍♂️

File: api/v1alpha1/clusterinstance_types.go (hunk 5 above)
Comment (sakhoury) [reply api/v1alpha1/clusterinstance_types.go:236]: Updated

File: api/v1alpha1/clusterinstance_types.go (hunk 6 above)
Comment (sakhoury) [reply api/v1alpha1/clusterinstance_types.go:446]: This should be small as we confine the permissible changes to the ClusterInstance spec fields.

File: api/v1alpha1/clusterinstance_types.go [hunk 7]
```diff
@@ -406,6 +448,27 @@ type ManifestReference struct {
 	Message string `json:"message,omitempty"`
 }
 
+type ReinstallHistory struct {
+	// Observed reinstall generation.
+	// +required
+	ObservedGeneration string `json:"observedGeneration"`
```
Comment (carbonin) [inline api/v1alpha1/clusterinstance_types.go:454]: Should this just be `generation` for the history?

Comment (sakhoury) [reply api/v1alpha1/clusterinstance_types.go:454]: yeah good point!

**Review by carbonin — APPROVED** [review]

**Pull request merged**